### PR TITLE
Validate spends of transparent coinbase outputs

### LINKED ISSUE #2329: State: Validate transparent coinbase outputs
## Motivation

Zebra needs to track the source of transparent coinbase UTXOs in the state, so we can validate:
- the coinbase maturity rule, and
- the mandatory transparent coinbase shielding rule.

(Zebra's state already ignores the genesis coinbase outputs.)

### Specifications

> A transaction with one or more transparent inputs from coinbase transactions MUST have no transparent outputs (i.e.tx_out_count MUST be 0). Inputs from coinbase transactions include Founders’ Reward outputs and funding stream outputs.

> A transaction MUST NOT spend a transparent output of a coinbase transaction from a block less than 100 blocks prior to the spend. Note that transparent outputs of coinbase transactions include Founders’ Reward outputs and transparent funding stream outputs.

> A transaction MUST NOT spend an output of the genesis block coinbase transaction. (There is one such zero-valued output, on each of Testnet and Mainnet.)

https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus

### Designs

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0004-asynchronous-script-verification.md, as edited in PR #1970:

Specifically:
* https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0004-asynchronous-script-verification.md#data-structures
* https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0004-asynchronous-script-verification.md#transparent-coinbase-consensus-rules
* Step 2 of https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0004-asynchronous-script-verification.md#lookup-implementation

## Related Work

Design RFC in #1970 
Transaction verifier in #2330

### LINKED ISSUE #2330: Transaction Verifier: Validate transparent coinbase outputs
## Motivation

Zebra needs to track the source of transparent coinbase UTXOs in the transaction verifier, so we can validate:
- the coinbase maturity rule, and
- the mandatory transparent coinbase shielding rule (this check might be skipped, check the final design).

### Specifications

> A transaction with one or more transparent inputs from coinbase transactions MUST have no transparent outputs (i.e.tx_out_count MUST be 0). Inputs from coinbase transactions include Founders’ Reward outputs and funding stream outputs.

> A transaction MUST NOT spend a transparent output of a coinbase transaction from a block less than 100 blocks prior to the spend. Note that transparent outputs of coinbase transactions include Founders’ Reward outputs and transparent funding stream outputs.

https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus

### Designs

See the design RFC from #1970

## Solution

Implement the Transaction Verifier parts of the design RFC in #1970:
  - _TODO: list tasks here once the RFC is completed_

## Related Work

Design RFC in #1970 
State in #2329

### LINKED ISSUE #2410: Improve `has_inputs_and_outputs` checks
In https://github.com/ZcashFoundation/zebra/pull/2398 we updated `has_inputs_and_outputs` to check for new consensus rules.

The check is a bit confusing now and even if we think it was good enough and have no major risks we will want to express the spec rules a bit better:

- https://github.com/ZcashFoundation/zebra/pull/2398#discussion_r659423340
- https://github.com/ZcashFoundation/zebra/pull/2398#pullrequestreview-693611614

This is a low priority cleanup task.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra needs to check the consensus rules that restrict how transparent outputs of coinbase transactions can be spent.

### Specifications

See #2329 and #2330.

### Designs

These designs are more complex than we actually need:

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0004-asynchronous-script-verification.md , particularly the changes in PR #1970 

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#requestawaitspendableutxo--outpoint-outpoint-spend_height-height-spend_restriction-spendrestriction-

## Solution

Implementation:
- Add a `CoinbaseSpendRestriction` enum and `Transaction` method
- Check UTXO spendability during contextual validation, using known valid UTXOs

Test fixes:
- Skip created UTXOs in the genesis block in arbitrary block chains
- Refactor out a new_transaction_ordered_outputs function
- Add Transaction::outputs_mut for tests
- Only spend valid UTXOs in arbitrary block chains

Closes #2329
Closes #2330
Closes #2410

### WIP Tests

Skip
  - [x] non-coinbase UTXOs
    - existing unit tests, acceptance tests, and cached state tests

Success
  - [x] coinbase UTXO spent to shielded outputs after 100 or more blocks 
    - existing unit tests, acceptance tests, and cached state tests
  - [x] coinbase UTXO spent to shielded outputs after exactly 100 blocks
    - modified existing state request test

Failure
  - [ ] coinbase UTXO spent to non-shielded outputs after 100 or more blocks
  - [x] coinbase UTXO spent to shielded outputs after less than 100 blocks

Test coverage
  - [x] make the proptest chain lengths slightly longer (ideally using a single constant)
  - [x] make sure that restriction-checked and unchecked proptest chains contain UTXO prevout spends

We could also do tests for finalized/non-finalized state, and both non-shielded and immature. But these tests are the essential ones.

I'm also running a full sync test locally on mainnet and testnet.

## Review

I'd like an initial review from @jvff while I'm working on the rest of the tests.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Move the design changes in PR #1970 to an "alternatives" section, and explain how the simpler design works.
(Or just revert most of PR #1970.)

---
This change is [Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)

### CODE REVIEW

File: zebra-state/src/request.rs
Comment (teor2345) [inline zebra-state/src/request.rs]:
```suggestion
    /// However, the spendability of a valid UTXO should not change, because
```

File: zebra-state/src/service/pending_utxos.rs
Comment (teor2345) [inline zebra-state/src/service/pending_utxos.rs]:
```suggestion
```

File: zebra-state/src/request.rs
Comment (daira) [inline zebra-state/src/request.rs]: The U in UTXO stands for Unspent, so they are potentially spendable by definition. I think this meant to say "transaction output" (twice). Also, I don't understand the last sentence; of course an unspent transaction output can potentially be spent, if a satisfying ScriptSig is known (or the chain rolled back to a point before a spent transaction output was spent).

**Review by jvff — COMMENTED** [review]
> Reviewed 10 of 14 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 12 of 14 files reviewed, 4 unresolved discussions (waiting on @teor2345)

___
*[zebra-chain/src/transparent/utxo.rs, line 84 at r3](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfXuHac6OHSqHiKsYj6:-MfXuHac6OHSqHiKsYj7:b-xy46pk) ([raw file](https://github.com/zcashfoundation/zebra/blob/35b5e5ec88bc3031734d4c8e56aa9219ec27abd6/zebra-chain/src/transparent/utxo.rs#L84)):*
> ```XML
>     SomeTransparentOutputs,
> 
>     /// The UTXO is spent in a transaction with all shielded outputs
> ```

Nit: maybe replace `all` with `only`?

```suggestion
    /// The UTXO is spent in a transaction with only shielded outputs.
```

File: zebra-state/src/error.rs
Comment (jvff) [inline zebra-state/src/error.rs]: Nit: Would it make sense to start breaking the long error messages in multi-line strings?

```suggestion
    #[error(
        "immature transparent coinbase spend: \
        attempt to spend {outpoint:?} at {spend_height:?}, \
        but spends are invalid before {min_spend_height:?}, \
        which is {MIN_TRANSPARENT_COINBASE_MATURITY:?} blocks \
        after it was created at {created_height:?}"
    )]
```

**Review by jvff — COMMENTED** [review]
> Reviewed 2 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: all files reviewed, 5 unresolved discussions (waiting on @teor2345)

File: zebra-state/src/service/check/utxo.rs
Comment (teor2345) [inline zebra-state/src/service/check/utxo.rs]: @jvff there are two alternatives here:
1. Just `validate_transparent_coinbase_spend` in the non-finalized state
2. Check it in both the non-finalized state and the UTXO requests 

We can `validate_transparent_coinbase_spend` in the non-finalized state, as long as we fix the proptests to produce valid spends.

If we do the validation here, we get rid of the state request changes in this PR. And we don't have to worry about UTXOs from invalid queued blocks making other invalid blocks pass this check. (Or making valid blocks fail.)

Or we do both checks, which gets rid of most invalid blocks early, and makes sure everything is valid. But it seems to be more complicated than we need.

Fixing the proptests should be doable. We'd need to delete all transparent inputs until we're past 100 blocks, then we substitute valid `OutPoint`s. We'll also need to make the tests use ~110 blocks, so we actually get good coverage. And check what we're doing with proptest genesis outputs.

File: zebra-state/src/request.rs
Comment (teor2345) [reply zebra-state/src/request.rs]: I agree that this wording is confusing, and so is the design. We're thinking about going with a simpler design. (And reverting this part of the change.)

The last sentence is meant to refer to the coinbase maturity and coinbase transparent output shielding consensus rules.

**Review by teor2345 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 7 of 14 files reviewed, 3 unresolved discussions (waiting on @daira, @jvff, and @teor2345)

___
*[zebra-chain/src/transparent/utxo.rs, line 84 at r3](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfXuHac6OHSqHiKsYj6:-MfacQF_1uNWkfK3-RPu:b-mdcimo) ([raw file](https://github.com/zcashfoundation/zebra/blob/6f03c9dd1e206343c3c91d012bd9866fa1bed05a/zebra-chain/src/transparent/utxo.rs#L84)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Nit: maybe replace `all` with `only`?

```suggestion
    /// The UTXO is spent in a transaction with only shielded outputs.
```
</blockquote></details>

Fixed in 4452d776c0e624e6501a77f55d1d99656dca2463
___
*[zebra-state/src/error.rs, line 104 at r3](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfVpsWKN9a64pJ9aTsg-r3-104:-MfafGAJ8k5MBMfvVR2o:b7njegx) ([raw file](https://github.com/zcashfoundation/zebra/blob/6f03c9dd1e206343c3c91d012bd9866fa1bed05a/zebra-state/src/error.rs#L104)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Nit: Would it make sense to start breaking the long error messages in multi-line strings?

```suggestion
    #[error(
        "immature transparent coinbase spend: \
        attempt to spend {outpoint:?} at {spend_height:?}, \
        but spends are invalid before {min_spend_height:?}, \
        which is {MIN_TRANSPARENT_COINBASE_MATURITY:?} blocks \
        after it was created at {created_height:?}"
    )]
```
</blockquote></details>

Fixed in ee0ea499bb6d8f08f9d4d251f838c80f4b3a7722

File: zebra-state/src/request.rs
Comment (teor2345) [reply zebra-state/src/request.rs]: This change was reverted as part of a rebase and force-push.

File: zebra-state/src/service/check/utxo.rs
Comment (teor2345) [reply zebra-state/src/service/check/utxo.rs]: I've updated the PR for option 1 - just validate transparent coinbase spends in the non-finalized state:
* 7a1ba4bc885db78bfdd280c0a44cd6d2d7a0f8e8 is the initial PR split
* 949479577b7f869e6efa3e80a0ec0c9cfdc3a063 does some comment fixes, function renames, and function moves
* e21515e045fd4a79838e54169dd479d01d409243 fix the proptest UTXOs
* e0cf1af3cdb9c065d431eb175c1fc69664b2040c ignore genesis coinbase outputs in proptests

File: zebra-state/src/service/check/utxo.rs
Comment (teor2345) [inline zebra-state/src/service/check/utxo.rs:81]: @oxarbitrage at this point we know we have the correct UTXOs, so we can validate `remaining_transaction_value` here.

File: zebra-chain/src/block/arbitrary.rs
Comment (teor2345) [inline zebra-chain/src/block/arbitrary.rs]: Once all the tests are working, I'll split this into a bunch of different methods

**Review by jvff — COMMENTED** [review]
> Reviewed 1 of 5 files at r6, 9 of 10 files at r7, 3 of 8 files at r8, 2 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 15 of 18 files reviewed, 5 unresolved discussions (waiting on @daira, @jvff, and @teor2345)

**Review by jvff — DISMISSED** [review]
> Reviewed 2 of 8 files at r8, 1 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: all files reviewed, 5 unresolved discussions (waiting on @daira and @teor2345)

**Review by teor2345 — COMMENTED** [review]
> Dismissed @daira from a discussion.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: all files reviewed, 3 unresolved discussions (waiting on @jvff and @teor2345)

**Review by teor2345 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 10 of 24 files reviewed, 1 unresolved discussion (waiting on @jvff and @teor2345)

___
*[zebra-chain/src/block/arbitrary.rs, line 368 at r9](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfbTsCxxNMHr680pT3v-r9-368:-MffjRB90oOx8A8vS78_:bm0xonb) ([raw file](https://github.com/zcashfoundation/zebra/blob/b1e8e275929e1b8e5d02bcd7e160fd575bed3873/zebra-chain/src/block/arbitrary.rs#L368)):*
<details><summary><i>Previously, teor2345 (teor) wrote…</i></summary><blockquote>

Once all the tests are working, I'll split this into a bunch of different methods
</blockquote></details>

Here are the new commits I added:

Test that generated chains contain at least one transparent spend 48e4b6b
Make generated chains long enough for reliable tests 0df08ca
Split chain generation into 3 functions da95c012 
Test that unshielded and immature transparent coinbase spends fail 69657822

This PR is getting a bit big. If you'd like, I can split these changes into other PRs, and get them merged first:

Refactor out a new_transaction_ordered_outputs function 8ead80c
Add Transaction::outputs_mut for tests 281fd68
Add transparent and shielded input and output methods to Transaction 75806738

The other changes will be a bit harder to split out.

**Review by jvff — COMMENTED** [review]
> Reviewed 1 of 9 files at r10, 2 of 7 files at r11, 3 of 5 files at r12.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 16 of 24 files reviewed, 2 unresolved discussions (waiting on @jvff and @teor2345)

___
*[zebra-state/src/service.rs, line 35 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-Mfj34I75hffyzRIEG9e:-Mfj34I75hffyzRIEG9f:b-v534bz) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service.rs#L35)):*
> ```XML
> 
> #[cfg(any(test, feature = "proptest-impl"))]
> pub mod arbitrary;
> ```

Just curious: is there a guideline for splitting module declarations? I want to know if/how I should them in my PRs :sweat_smile:

**Review by jvff — COMMENTED** [review]
> Reviewed 3 of 9 files at r10, 1 of 7 files at r11, 2 of 5 files at r12.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 22 of 24 files reviewed, 5 unresolved discussions (waiting on @jvff and @teor2345)

___
*[zebra-chain/src/block/arbitrary.rs, line 1 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-Mfj4Ejr9MfIpaZuHeNn:-Mfj4Ejr9MfIpaZuHeNo:bqelij5) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-chain/src/block/arbitrary.rs#L1)):*
> ```XML
> //! Randomised property testing for [`Block`]s
> ```

Nit: Maybe add a period at the end?
___
*[zebra-state/src/service/arbitrary.rs, line 21 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-Mfj3pDx8psBiIOzqIGm:-Mfj3pDx8psBiIOzqIGn:b-fnbved) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service/arbitrary.rs#L21)):*
> ```XML
> use super::*;
> 
> /// The minimum height required for reliable non-finalized state property tests.
> ```

Minor: Should the documentation say something like `maximum height for unreliable...`? Also wondering if it would make sense to add a margin of error to it :thinking: 
___
*[zebra-state/src/service/check/tests/utxo.rs, line 76 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfjVHRC9znwaUFOKb9D:-MfjVHRC9znwaUFOKb9E:b-u4j9wj) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service/check/tests/utxo.rs#L76)):*
> ```XML
> 
>     let min_spend_height = Height(created_height.0 + MIN_TRANSPARENT_COINBASE_MATURITY);
>     let spend_height = Height(min_spend_height.0 - 1);
> ```

Minor: Should we add a test where `spend_height = min_spend_height`?

**Review by jvff — DISMISSED** [review]
:lgtm:

> Reviewed 1 of 9 files at r10, 1 of 7 files at r11.
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-:-MfjXXAG9qEoXBN-QZIM:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @teor2345)

___
*[zebra-chain/src/block/arbitrary.rs, line 368 at r9](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfbTsCxxNMHr680pT3v-r9-368:-MfjW0Kq-aG9fPmfIXGp:bvojfp8) ([raw file](https://github.com/zcashfoundation/zebra/blob/b1e8e275929e1b8e5d02bcd7e160fd575bed3873/zebra-chain/src/block/arbitrary.rs#L368)):*
<details><summary><i>Previously, teor2345 (teor) wrote…</i></summary><blockquote>

Here are the new commits I added:

Test that generated chains contain at least one transparent spend 48e4b6b
Make generated chains long enough for reliable tests 0df08ca
Split chain generation into 3 functions da95c012 
Test that unshielded and immature transparent coinbase spends fail 69657822

This PR is getting a bit big. If you'd like, I can split these changes into other PRs, and get them merged first:

Refactor out a new_transaction_ordered_outputs function 8ead80c
Add Transaction::outputs_mut for tests 281fd68
Add transparent and shielded input and output methods to Transaction 75806738

The other changes will be a bit harder to split out.
</blockquote></details>

I think that won't be necessary. Most of it has been reviewed :grin:

File: zebra-chain/src/block/arbitrary.rs
Comment (teor2345) [inline zebra-chain/src/block/arbitrary.rs]:
```suggestion
//! Randomised property testing for [`Block`]s.
```

File: zebra-state/src/service/arbitrary.rs
Comment (teor2345) [inline zebra-state/src/service/arbitrary.rs]:
```suggestion
/// The chain length for state proptests.
///
/// Shorter lengths increase the probability of proptest failures.
```

File: zebra-chain/src/block/arbitrary.rs
Comment (teor2345) [inline zebra-chain/src/block/arbitrary.rs]:
```suggestion
/// proptest_cases = 256
/// number_of_proptests = 5 as of July 2021 (PREVOUTS_CHAIN_HEIGHT and PartialChain tests)
/// shielded_input^(expected_transactions * expected_inputs * PREVOUTS_CHAIN_HEIGHT) * proptest_cases * number_of_proptests
```

**Review by teor2345 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 22 of 24 files reviewed, 2 unresolved discussions (waiting on @jvff)

___
*[zebra-state/src/service.rs, line 35 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-Mfj34I75hffyzRIEG9e:-MfjiH2p0XA9LEzzw3OM:b-pmo52a) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service.rs#L35)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Just curious: is there a guideline for splitting module declarations? I want to know if/how I should them in my PRs :sweat_smile: 
</blockquote></details>

I have no idea :shrug: 

This might be a bad habit of mine from C, where conditional compilation happens using macros in `#if ... #endif` blocks.

Happy to go with a Rust guideline if we find one.
___
*[zebra-state/src/service/arbitrary.rs, line 21 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-Mfj3pDx8psBiIOzqIGm:-MfjlHT_2DlIWJv8-dLm:bl5j0jy) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service/arbitrary.rs#L21)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Minor: Should the documentation say something like `maximum height for unreliable...`? Also wondering if it would make sense to add a margin of error to it :thinking: 
</blockquote></details>

I clarified the comment here, and for `PREVOUTS_CHAIN_HEIGHT`.

There is already a margin of error built in to `PREVOUTS_CHAIN_HEIGHT`, currently the failure rate should be around 1 in 10 million test runs.

I just picked a really rare probability, we can tweak it if it becomes an issue later on.
___
*[zebra-state/src/service/check/tests/utxo.rs, line 76 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfjVHRC9znwaUFOKb9D:-MfjnOe7AdNxTPivEsfB:b-it0now) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service/check/tests/utxo.rs#L76)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Minor: Should we add a test where `spend_height = min_spend_height`?
</blockquote></details>

Good idea, I had one of those, but it might have got deleted when I simplified the PR.

Also I think the chain itself checks this, but it's an easy test to write, so I'll do it now.

**Review by teor2345 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2525)* status: 21 of 24 files reviewed, all discussions resolved (waiting on @jvff)

___
*[zebra-state/src/service/check/tests/utxo.rs, line 76 at r12](https://reviewable.io/reviews/zcashfoundation/zebra/2525#-MfjVHRC9znwaUFOKb9D:-MfjolSI5ZpqBKLgE-PK:blg5zwh) ([raw file](https://github.com/zcashfoundation/zebra/blob/6965782265bbb51216d4c28fcd76215f6da34428/zebra-state/src/service/check/tests/utxo.rs#L76)):*
<details><summary><i>Previously, teor2345 (teor) wrote…</i></summary><blockquote>

Good idea, I had one of those, but it might have got deleted when I simplified the PR.

Also I think the chain itself checks this, but it's an easy test to write, so I'll do it now.
</blockquote></details>

Done in b11310f5

**Comment by teor2345** [comment]
I just worked out that this PR is blocking the next PR after #2486.
There's a chain of about 5 PRs we need to do there.

So I might just admin-merge this PR after CI passes.
After the last approval, I added one requested test and some comments.

**Comment by teor2345** [comment]
Admin-merged, because the PR was approved before minor changes, and it's blocking other work

**Pull request merged**
